plugins/parsers/avro/parser.go

```diff
@@ -0,0 +1,238 @@
+package avro
+
+import (
+	"encoding/binary"
+	"encoding/json"
+	"fmt"
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/internal"
+	"github.com/influxdata/telegraf/metric"
+	"github.com/influxdata/telegraf/plugins/parsers"
+	"github.com/jeremywohl/flatten"
+	"github.com/linkedin/goavro/v2"
+	"time"
+)
+
+type Parser struct {
+	MetricName      string   `toml:"metric_name"`
+	SchemaRegistry  string   `toml:"avro_schema_registry"`
+	Schema          string   `toml:"avro_schema"`
+	Measurement     string   `toml:"avro_measurement"`
+	Tags            []string `toml:"avro_tags"`
+	Fields          []string `toml:"avro_fields"`
+	Timestamp       string   `toml:"avro_timestamp"`
+	TimestampFormat string   `toml:"avro_timestamp_format"`
+	DiscardArrays   bool     `toml:"avro_discard_arrays"`
+	FieldSeparator  string   `toml:"avro_field_separator"`
+	DefaultTags     map[string]string
+	TimeFunc        func() time.Time
+
+	Log telegraf.Logger `toml:"-"`
+}
+
+func (p *Parser) Parse(buf []byte) ([]telegraf.Metric, error) {
+	var schema string
+
+	if len(buf) < 5 {
+		err := fmt.Errorf("buf is %d bytes; must be at least 5", len(buf))
+		return nil, err
+	}
+	schemaID := int(binary.BigEndian.Uint32(buf[1:5]))
+	binaryData := buf[5:]
+
+	switch {
+	case p.SchemaRegistry != "":
+		schemaRegistry := NewSchemaRegistry(p.SchemaRegistry)
+		retrSchema, err := schemaRegistry.getSchema(schemaID)
+		if err != nil {
+			return nil, err
+		}
+		schema = retrSchema
+	default:
+		schema = p.Schema
+	}
+
+	codec, err := goavro.NewCodec(schema)
+
+	if err != nil {
+		return nil, err
+	}
+
+	native, _, err := codec.NativeFromBinary(binaryData)
+	if err != nil {
+		return nil, err
+	}
+
+	m, err := p.createMetric(native, schema)
+	if err != nil {
+		return nil, err
+	}
+
+	return []telegraf.Metric{m}, nil
+}
+
+func (p *Parser) ParseLine(line string) (telegraf.Metric, error) {
+	metrics, err := p.Parse([]byte(line))
+	if err != nil {
+		return nil, err
+	}
+
+	if len(metrics) != 1 {
+		return nil, fmt.Errorf("line contains multiple metrics")
+	}
+
+	return metrics[0], nil
+}
+
+func (p *Parser) SetDefaultTags(tags map[string]string) {
+	p.DefaultTags = tags
+}
+
+func (p *Parser) parseTimestamp(timestamp interface{}) (time.Time, error) {
+	if timestamp == nil {
+		return p.TimeFunc(), nil
+	}
+
+	if p.TimestampFormat == "" {
+		return p.TimeFunc(), fmt.Errorf("must specify timestamp format")
+	}
+
+	metricTime, err := internal.ParseTimestamp(p.TimestampFormat, timestamp, "UTC")
+	if err != nil {
+		return p.TimeFunc(), err
+	}
+
+	return metricTime, nil
+}
+
+func (p *Parser) createMetric(native interface{}, schema string) (telegraf.Metric, error) {
```

Comment:
Is there always only one metric in the data or is it possible to send "arrays" of metrics?

Reply:
One metric per message, at least according to the README.md I inherited (and I have not seen any behavior to suggest otherwise).